head.go

```diff
@@ -752,22 +762,89 @@ func (h *Head) Delete(mint, maxt int64, ms ...labels.Matcher) error {
 		}
 		// Delete only until the current values and not beyond.
 		t0, t1 = clampInterval(mint, maxt, t0, t1)
-		stones = append(stones, Stone{p.At(), Intervals{{t0, t1}}})
+		if err := h.delete(p.At(), Intervals{{t0, t1}}); err != nil {
+			return errors.Wrap(err, "delete samples")
+		}
+		dirty = true
 	}
-
 	if p.Err() != nil {
 		return p.Err()
 	}
-	var enc RecordEncoder
+	if dirty {
+		h.gc()
+	}
```

Comment:
The symbols and values gets updated. So if some labels no longer exist after deleting, then it would be removed here.
It's tested here: https://github.com/prometheus/tsdb/blob/158f3c19bb5c7671c3ff18c225dd1d9bbb82c0e6/head_test.go#L400-L404